server/src/main/java/org/opensearch/common/settings/WriteableSetting.java

```diff
@@ -0,0 +1,275 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.settings;
+
+import org.opensearch.Version;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.common.io.stream.Writeable;
+import org.opensearch.common.settings.Setting.Property;
+import org.opensearch.common.unit.ByteSizeValue;
+import org.opensearch.common.unit.TimeValue;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.EnumSet;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Wrapper for {@link Setting} with {@link #writeTo(StreamOutput)} implementation dependent on the setting type.
+ *
+ * @opensearch.internal
+ */
+public class WriteableSetting implements Writeable {
+
+    /**
+     * The Generic Types which this class can serialize.
+     */
+    public enum SettingType {
+        Boolean,
+        Integer,
+        Long,
+        Float,
+        Double,
+        String,
+        TimeValue, // long + TimeUnit
+        ByteSizeValue, // long + ByteSizeUnit
+        Version
```

Comment:
It seems like we missed `Settings` (aka grouped) and `List<?>`? On a general note, may be we should redesign this part and have each `Setting<?>` implementing `Writeable`, here is why:
 - the `Setting<?>` could be created only using factory methods, at this point it is crystal clear what is the type of the setting
 -  since each `Setting<?>` knows its type, it could implement `Writeable` easily

Reply:
@dbwiddis you want to take a look here?

Reply:
Not missed: just a TODO.

There are open PRs for list settings: https://github.com/opensearch-project/opensearch-sdk-java/issues/155
and group settings: https://github.com/opensearch-project/opensearch-sdk-java/issues/156

There's also a need to handle the functional interfaces that implement settings validation (e.g., a minimum or maximum value): https://github.com/opensearch-project/opensearch-sdk-java/issues/154

Meta:  https://github.com/opensearch-project/opensearch-sdk-java/issues/148

Reply:
> On a general note, may be we should redesign this part and have each Setting<?> implementing `Writeable`

I agree in principle.  The `WriteableSetting` class was an attempt to work around that without modifying the existing code, but essentially implements the same functionality.  But that still creates problems with the functional interfaces that can not be easily serialized.

Reply:
> the Setting<?> could be created only using factory methods

This would be a reduction in functionality from the existing implementation, see: https://github.com/opensearch-project/OpenSearch/blob/main/plugins/examples/custom-settings/src/main/java/org/opensearch/example/customsettings/ExampleCustomSettingsConfig.java

The META issue in the above comment is basically an attempt to recreate the functionality of this existing plugin.

Reply:
Sounds good, thank @dbwiddis